cli/lsp/config.rs

```diff
@@ -1427,181 +1423,193 @@ impl ConfigData {
 
 #[derive(Debug, Default)]
 pub struct ConfigTree {
-  root: Mutex<Option<(ModuleSpecifier, Arc<ConfigData>)>>,
+  first_folder: Mutex<Option<ModuleSpecifier>>,
+  scopes: Mutex<BTreeMap<ModuleSpecifier, Arc<ConfigData>>>,
 }
 
 impl ConfigTree {
   pub fn root_data(&self) -> Option<Arc<ConfigData>> {
-    self.root.lock().as_ref().map(|(_, d)| d.clone())
-  }
-
-  pub fn root_config_file(&self) -> Option<Arc<ConfigFile>> {
     self
-      .root
+      .first_folder
       .lock()
       .as_ref()
-      .and_then(|(_, d)| d.config_file.clone())
+      .and_then(|s| self.scopes.lock().get(s).cloned())
   }
 
   pub fn root_ts_config(&self) -> Arc<TsConfig> {
     self
-      .root
+      .first_folder
       .lock()
       .as_ref()
-      .map(|(_, d)| d.ts_config.clone())
+      .and_then(|s| self.scopes.lock().get(s).map(|d| d.ts_config.clone()))
       .unwrap_or_else(|| Arc::new(default_ts_config()))
   }
 
   pub fn root_vendor_dir(&self) -> Option<PathBuf> {
-    self
-      .root
-      .lock()
-      .as_ref()
-      .and_then(|(_, d)| d.vendor_dir.clone())
+    self.first_folder.lock().as_ref().and_then(|s| {
+      self.scopes.lock().get(s).and_then(|d| d.vendor_dir.clone())
+    })
   }
 
   pub fn root_lockfile(&self) -> Option<Arc<Mutex<Lockfile>>> {
     self
-      .root
+      .first_folder
       .lock()
       .as_ref()
-      .and_then(|(_, d)| d.lockfile.clone())
+      .and_then(|s| self.scopes.lock().get(s).and_then(|d| d.lockfile.clone()))
+  }
+
+  pub fn root_import_map(&self) -> Option<Arc<ImportMap>> {
+    self.first_folder.lock().as_ref().and_then(|s| {
+      self.scopes.lock().get(s).and_then(|d| d.import_map.clone())
+    })
```

Comment:
We're acquiring locks in different orders, ex. `first_folder.lock()` here then `self.scopes.lock()` and then below on line 1472 it acquires `self.scopes.lock()` and then `self.first_folder.lock()`. Not sure if it's related, but that might be the cause of the hanging.

This use of multiple locks is very sketchy and we should remove it. Looking at the code a bit more, it looks like we could remove these locks entirely and then create a `ConfigTreeSnapshot` to send to tsc. Sending mutable data in the tsc snapshot is not ideal.

Reply:
Turns out we don't need locks at all since we can just update with a mut reference...